src/HtmlUtils.tsx

```diff
@@ -639,6 +642,51 @@ export function linkifyAndSanitizeHtml(dirtyHtml: string, options = linkifyMatri
     return sanitizeHtml(linkifyString(dirtyHtml, options), sanitizeHtmlParams);
 }
 
+/**
+ * Recurses depth-first through a DOM tree, adding tooltip previews for link elements.
+ *
+ * @param {Element[]} rootNodes - a list of sibling DOM nodes to traverse to try
+ *   to add tooltips.
+ * @param {Element[]} ignoredNodes: a list of nodes to not recurse into.
+ */
+export function tooltipifyLinks(rootNodes: ArrayLike<Element>, ignoredNodes: Element[]) {
+    if (!PlatformPeg.get()?.needsUrlTooltips()) {
+        return;
+    }
+
+    let node = rootNodes[0];
+
+    while (node) {
+        let tooltipified = false;
+
+        if (ignoredNodes.indexOf(node) >= 0) {
+            node = node.nextSibling as Element;
+            continue;
+        }
+
+        if (node.tagName === "A" && node.getAttribute("href")
+            && node.getAttribute("href") !== node.textContent.trim()
+        ) {
+            const container = document.createElement('span');
+            const href = node.getAttribute("href");
+            // Disable focusing on the tooltip target to avoid double / nested focus. The contained anchor element
+            // itself allows focusing which also triggers the tooltip.
+            const tooltip = <TextWithTooltip tabIndex={-1} tooltip={new URL(href, window.location.href).toString()}>
+                <span dangerouslySetInnerHTML={{ __html: node.outerHTML }} />
+            </TextWithTooltip>;
+            ReactDOM.render(tooltip, container);
```

Comment:
Don't we also need to unmount this to clean up the shadow dom?

Reply:
Yes, I think we do and I believe the fact that I'm not currently tracking the created tooltip components is also the reason for the triple-focusing that Jano observed. Sorry, I didn't get to make those changes yet. Should've been more explicit about this not being finished yet.

Reply:
Please use the draft state for WIP stuff

Reply:
> I apologize if me pushing part of the fixes for the earlier review has made this look as if it's ready to be re-reviewed.

Nope, my fault entirely, I saw high-risk a11y regressions and I stepped on this, sorry

Reply:
The risk is real so I very much appreciate the review in general! 🙂

Reply:
Unmount is happening now